samples/96b_argonkey: divided test into sensor and microphone

avisconti opened the pull request:
We have now two code samples available for ArgonKey board:

    1. sensors sample, fetching data from accel/gyro/mag,
       barometer, humidity and proximity sensors

    2. microphone sample, acquiring 3s audio from the
       on-board microphone.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@erwango @galak

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+the back of the board on P3.5 (TX) and P3.7 (RX).
+
+As soon as the acquisition starts the green led glows. At the end of the
+acquisition both the green and red led goes on and the PCM stream is sent
```
```suggestion
acquisition both the green and red LEDs go on and the PCM stream is sent
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+A USB to TTL 1V8 serial cable may be attached to the low speed connector on
+the back of the board on P3.5 (TX) and P3.7 (RX).
+
+As soon as the acquisition starts the green led glows. At the end of the
```
```suggestion
As soon as the acquisition starts the green LED glows. At the end of the
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+- acquisition ends:       GRN on  - RED on
+- output ends:            GRN off - RED on
+
+The characteristic of the PCM audio are hardcoded in the example:
```
```suggestion
The characteristics of the PCM audio are hardcoded in the example:
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+
+The characteristic of the PCM audio are hardcoded in the example:
+
+- 32KHz of sample rate
```
```suggestion
- 32KHz sample rate
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+- 16 bits per sample
+- 1 channel (mono), as only 1 microphone is available
+
+3 seconds of acquisition @32Kz of sampling rate corresponds to 96000 sample
```
```suggestion
Three seconds of acquisition at a 32KHz sampling rate yields 96,000 16-bit samples.
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+
+Since the MP34DT05 microphone is PDM, we need to apply an oversampling factor,
+which is 64 in this example, resulting in a 2MHz bit clock. During the PCM
+conversion the OpenPDMFilter library is filtering out and decimating the data
```
```suggestion
conversion, the OpenPDMFilter library filters out and decimates the data
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+by the same 64 factor.
+
+At the end of the acquisition the PCM data will be printed on the terminal
+emulator in either binary or ascii format. The output is controlled by
```
```suggestion
emulator in either binary or ASCII format. The output is controlled by
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+
+At the end of the acquisition the PCM data will be printed on the terminal
+emulator in either binary or ascii format. The output is controlled by
+PCM_OUTPUT_IN_ASCII definition, which is off by default.
```
mention where this define is declared (samples/boards/96b_argonkey/microphone/src/main.c)

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+Binary PCM Output
+-----------------
+
+The ttyUSB0 port must be configured in raw mode in order to avoid to have
```
```suggestion
The ttyUSB0 port must be configured in raw mode to avoid having
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+The ttyUSB0 port must be configured in raw mode in order to avoid to have
+characters 'cooked' out.
+
+ .. code-block:: console
```
This code block (and contents) are indented by one extra space.  (The ``.. code-block`` directive should start in column 1.

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+    stty -F /dev/ttyUSB0 115200 raw
+    cat /dev/ttyUSB0 > /tmp/sound.raw
+
+NOTE!
```
Change this to:

    ..note:: In case the character 0x0a is interpreted as NL and an 0x0d (CR) is added,
       you may need to remove it::

          dos2unix -f /tmp/sound.raw

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+----------------
+
+It is also possible to recompile and to have PCM output in ASCII, which needs
+to be converted in binary later on. The output format is the following:
```
```suggestion
to be converted to binary later on. The output format is the following:
```

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+Play PCM Audio
+--------------
+
+At this point in either case we ended up with a binary PCM file, say sound.raw.
```
Change this to:

    Now that we have a binary PCM file (say sound.raw), you can use,
    for example, the audicity open source editor/player to load and play it.

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+The audacity open source audio editor/player can be used to load and play the
+PCM audio file.
+
+Just use the 'Import->Raw Data' item from menu and load the sound.raw as
```
```suggestion
Use the 'Import->Raw Data' menu to load the sound.raw file as
```

dbkinder commented on samples/boards/96b_argonkey/sensors/README.rst:
```diff
@@ -12,9 +12,15 @@ It makes use of both the trigger and poll methods.
 ************
 
-This sample just requires the ArgonKey board.
+This sample just requires the ArgonKey board. The board can be fed
```
You mention later you'll need a USB to TTL 1V8 serial cable.

Also, by "fed" do you mean "powered"?

dbkinder commented on samples/boards/96b_argonkey/sensors/README.rst:
```diff
@@ -12,9 +12,15 @@ It makes use of both the trigger and poll methods.
+
+- mezzanine mode, plugging the ArgonKey to HiKey board thru its 96Board
+  low-speed connector
+- standalone mode, feeding directly 5V on P1 connector
```
should "feeding"  be "powering"?

dbkinder commented on samples/boards/96b_argonkey/sensors/README.rst:
```diff
@@ -49,27 +55,40 @@ Building and Running
 
+A USB to TTL 1V8 serial cable may be attached to the low speed connector on
+the back of the board on P3.5 (TX) and P3.7 (RX). User may use a simple
+terminal emulator like minicom to capture the output.
```
```suggestion
terminal emulator, such as minicom, to capture the output.
```

erwango commented on samples/boards/96b_argonkey/microphone/src/OpenPDMFilter.c:
```diff
@@ -0,0 +1,331 @@
+/*
+ * @file    OpenPDMFilter.c
```
From my understanding, this should be made available as a lib (such as ext/hal/st/lib/sensor/vl53l0x/) and accessed through a specific driver (cf: drivers/sensor/vl53l0x), then used from this sample.
This should be done in several commits and the commit for lib should provide more info (can't find the link in zephyr doc, here is an example):
```
    title
    
   body    

    Origin: ST Microelectronics
    License: ..
    URL: where to get it
    Commit: (or version)
    Purpose: 
    Maintained-by: External
    
    Signed-off-by: ....
```

avisconti replied:
OK, I agree to move the OpenPDMFilter library under ext/hal/st/lib.

I'm wondering if the microphone driver is really needed. I was more thinking on
using the lib API in the application itself. A driver is maybe too much, as it is
just a passive element that starts sending a PCM stream as soon as it is powered and clocked.
It does not have registers to configure it.

dbkinder replied:
Since this is not Apache 2.0 licensed code, check out https://docs.zephyrproject.org/latest/contribute/contribute_guidelines.html#contributing-non-apache-2-0-licensed-components
@nashif @galak

avisconti replied:
Thanks @dbkinder 
I had already discussed this point with @erwango in pvt, but this link is pretty useful!

avisconti replied:
@dbkinder  @erwango 
I had modified the OpenPDMFilter library in order to be compliant with zephyr coding rules, but now I have this opinion to revert it back to the original version from Mbed os. I guess this is the right way to do it.

erwango commented on samples/boards/96b_argonkey/microphone/src/main.c:
```diff
@@ -0,0 +1,224 @@
+#include "OpenPDMFilter.h"
+#endif
+
+#ifdef CONFIG_I2S
```
You should use this flag here, sample has no meaning if not enabled. Rather we should add CONFIG_I2S in prj.conf if not done already.

avisconti replied:
Yes, I added it in prj.conf for microphone sample.

avisconti replied:
Maybe I could completely remove CONFIG_I2S in main.c.... I agree that it does not have much sense.

erwango commented on samples/boards/96b_argonkey/microphone/src/main.c:
```diff
@@ -0,0 +1,224 @@
+	struct device *adev = device_get_binding(CONFIG_I2S_5_NAME);
+
+	if (!adev) {
+		printk("Could not get pointer to %s sensor\n",
```
s/sensor/device

avisconti replied:
ACK

erwango commented on samples/boards/96b_argonkey/microphone/src/main.c:
```diff
@@ -0,0 +1,224 @@
+	filter_init.Decimation = OVERSAMPLING_FACTOR;
+	filter_init.MaxVolume = 64;
+
+	Open_PDM_Filter_Init(&filter_init);
```
Access should be done through a driver.

avisconti replied:
But I agree to move the OpenPDMFilter library in the ext/hal/st/lib directory to be used by
everyone. Please not that this library is the one with following license:

http://www.st.com/software_license_agreement_image_v2

erwango replied:
I think that a empty driver would be required just for the library license, it is provided to work only with MP34DT05, it could and should be enforced by the use of a dedicated driver.

avisconti replied:
OK, I'm leaving for the day. I'm resubmitting a new PR with everything addresses by you and dbkinder except this stuff. I will work on it tomorrow morning.

erwango commented on samples/boards/96b_argonkey/microphone/src/main.c:
```diff
@@ -0,0 +1,224 @@
+				     HTONS(((uint16_t *)rx_block)[i]);
+		}
+
+		Open_PDM_Filter_64((u8_t *) rx_block,
```
Access to be done through a driver

erwango commented on samples/boards/96b_argonkey/microphone/src/main.c:
```diff
@@ -0,0 +1,224 @@
+	signal_sampling_stopped();
+
+	/* print PCM stream */
+	#ifdef PCM_OUTPUT_IN_ASCII
```
#ifdef should not be indented

erwango commented on samples/boards/96b_argonkey/microphone/src/main.c:
```diff
@@ -0,0 +1,224 @@
+		k_mem_slab_free(&rx_mem_slab, &rx_block);
+
+		if (ms_count++ >= NUM_MS)
+			break;
```
Use { }

avisconti replied:
I always forget...

erwango commented on samples/boards/96b_argonkey/microphone/src/main.c:
```diff
@@ -0,0 +1,224 @@
+	Open_PDM_Filter_Init(&filter_init);
+
+	for (i = 0; i < 32*NUM_MS; i++)
+		pcm_out[i] = 0x0;
```
Use { }

erwango reviewed (changes requested):
Good application, but you should split part of this into a microphone driver and a lib in dedicated folder.

avisconti commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+Requirements
+************
+
+This sample just requires the ArgonKey board. The board can be fed
```
I'll put 'powered' instead of 'fed' globally.
I'll add also the cable requirement.

avisconti commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,131 @@
+
+- mezzanine mode, plugging the ArgonKey to HiKey board thru its 96Board
+  low-speed connector
+- standalone mode, feeding directly 5V on P1 connector
```
Yes.

avisconti replied:
Well, actually it doesn't sound good to me to repeat the word 'power'.
I can probably say:

"The board can be powered [.......] PROVIDING 5V directly on P1 connector."

What do you think?

avisconti replied:
Maybe SUPPLYING ?

dbkinder commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,127 @@
+Building and Running
+********************
+
+ .. zephyr-app-commands::
```
Sorry, missed this in the previous review.  Directives, such as ``.. zephyr-app-command`` and ``.. code-block`` should be indented at the same level as the paragraphs around them.  For the directive here on line 33, there should be no leading spaces.

avisconti replied:
ok

dbkinder commented on samples/boards/96b_argonkey/sensors/README.rst:
```diff
@@ -49,27 +55,40 @@ Building and Running
+A USB to TTL 1V8 serial cable may be attached to the low speed connector on
+the back of the board on P3.5 (TX) and P3.7 (RX). User may use a simple
+terminal emulator, such as minicom, to capture the output.
+
```
The ``.. code-block`` directive below should not be indented by a space.  Can you also fix the indent on the code-block on line 23.

dbkinder commented on samples/boards/96b_argonkey/sensors/README.rst:
```diff
@@ -40,7 +46,7 @@ Building and Running
 
  .. zephyr-app-commands::
-    :zephyr-app: samples/boards/96b_argonkey
+    :zephyr-app: samples/boards/96b_argonkey/sensors
```
The ``.. zephyr-app-command`` directive above should not be indented by a space.

dbkinder reviewed (changes requested):
Thanks for the previous doc fixes.  I added some notes about indenting directives (such as ``code-block``).

avisconti committed:
ext: Added OpenPDM2PCM ST external library

This library, written by STMicroelectronics, is used to convert an
audio stream from PDM format to PCM format through a signal filtering
and decimation.

Library is located in ext/hal/st/lib/audio.

Origin: ST Microelectronics
License: Apache 2.0
URL: https://os.mbed.com/
Commit: 25:f2c04f757003
Purpose: reconstruct the audio signal produced by ST MEMS microphone
Maintained-by: External

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
drivers/audio: Add support to ST digital microphones (MPxxDTyy)

This driver supports ST Microelectronics digital pdm microphones
(MPxxDTyy) connected through different peripherals. Currently only
I2S is supported.

The driver makes use internally of the OpenPDM2PCM library
to convert the PDM audio stream to PCM. Currently the
oversampling factor is fixed to 64.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
board/argonkey: Add MP34DT05 microphone in DTS

Add MP34DT05 pdm microphone in DTS under i2s_5 node.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
samples/96b_argonkey: divided test into sensor and microphone

We have now two code samples available for ArgonKey board:

    1. sensors sample, fetching data from accel/gyro/mag,
       barometer, humidity and proximity sensors

    2. microphone sample, acquiring 5s audio @16KHz from the
       on-board microphone.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
dts/bindings: audio: add binding for ST MPXXDYYY pdm microphones on I2S

Provide dts yaml bindings for ST MPXXDYYY microphones family on I2S.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
I have resubmitted the PR taking care of following things:

1. I moved the OpenPDMFilter library in ext/hal/st/lib/audio
    I prepared a README file with the library description including the license.
    The official library name is OpenPDM2PCM

2. I prepared a MP34DT05 microphone driver based on the DMIC i/f explained above.
    This driver uses of the openPDMFilter lib for the PDM 2 PCM conversion

3. I changed the ArgonKey board code to introduce the MP34DT05 microphone
     information in DTS. To be able to do this work I also had to introduce the dts
     bindings for the generic mp32dt05 I2S device (dts/bindings/audio/st,mp34dt05.yaml)

4. I restructured the sample code for ArgonKey.
    Now it has two samples:
        microphone sample, using dmic APIs
         sensor sample, to test sensor part (same as before)

Since all this work is inter-dependent I prefer to keep everything inside the same PR.

avisconti commented:
Fixed few things in the PR. Plus I re-order the patchset in something more logic.

@erwango 
Everything is now in the way you requested. Please take a look to the OPenPDM2PCM library as everything else depends by this one.

@dbkinder 
I reported all changes requested by you, but I also added few more info. Can you possibly review it again?

erwango commented on drivers/audio/mp34dt05.c:
```diff
@@ -0,0 +1,210 @@
+	struct mp34dt05_driver_config *const dev_cfg = DEV_CFG(dev);
+
+	/* bind I2S */
+	dev_cfg->i2s_device = device_get_binding(DT_I2S_5_NAME);
```
You should not use directly "DT_I2S_5_NAME" here.
You should make this components a i2s-device on the model of spi-device and then here use generated DT_XXXX_BUS_NAME

avisconti replied:
Actually I do not see for I2S a BUS_NAME property as there is for SPI. I only see:

#define DT_I2S_5_BASE_ADDRESS      DT_ST_STM32_I2S_40015000_BASE_ADDRESS                                                              
#define DT_I2S_5_IRQ_PRI                    DT_ST_STM32_I2S_40015000_IRQ_0_PRIORITY
#define DT_I2S_5_NAME                        DT_ST_STM32_I2S_40015000_LABEL
#define DT_I2S_5_IRQ                            DT_ST_STM32_I2S_40015000_IRQ_0
#define DT_I2S_5_CLOCK_BITS            DT_ST_STM32_I2S_40015000_CLOCK_BITS
#define DT_I2S_5_CLOCK_BUS             DT_ST_STM32_I2S_40015000_CLOCK_BUS

Is it possible that this is due to the fact that I2S dts, differently from SPI, does not use #address-cells and #size-cells clause? See @galak comments on #9034.

erwango replied:
We're actually looking for a DT_XXXX_MASTER_DEV_NAME, cf other comment

erwango commented on boards/arm/96b_argonkey/96b_argonkey.dts:
```diff
@@ -70,6 +70,11 @@
 	status = "ok";
+
+	mp34dt05 {
+		compatible = "st,mp34dt05";
```
Provide yaml file matching this compatible.
It should be instanciated as a i2s-device, which rely on a i2s bus (should be somehow similar to spi / spi-device model).
Then, you'll get a DT_MP34DT05_BUS_NAME or simialr you'll be able to use in the driver

avisconti replied:
I think I did. See 7939a51. The system generates infact:

#define DT_ST_MP34DT05_40015000_MP34DT05_LABEL             "MP34DT05"
#define DT_MP34DT05_DEV_NAME       DT_ST_MP34DT05_40015000_MP34DT05_LABEL

Isn't it correct?

erwango replied:
@avisconti ,

> I think I did

Indeed, i overlooked that part, my bad.
You then need to establish this is a device of the i2s bus, cf next comment

avisconti commented on ext/hal/st/lib/CMakeLists.txt:
```diff
@@ -17,5 +17,6 @@ if(CONFIG_HAS_STLIB)
     zephyr_sources(sensor/vl53l0x/api/core/src/vl53l0x_api_core.c)
     zephyr_sources(sensor/vl53l0x/api/core/src/vl53l0x_api_strings.c)
   endif()
+    zephyr_include_directories(audio/microphone)
```
sure! It makes sense.

erwango commented on dts/bindings/audio/st,mp34dt05.yaml:
```diff
@@ -0,0 +1,25 @@
+---
+title: ST Microelectronics MP34DT05 pdm microphone
+version: 0.1
+
```
Add here:
```
parent:
    bus: i2s
```
Then, in i2s.yaml, add the following:
```
child:
    bus: i2s
```
Then, you'll get a DT_MP34DT05_MASTER_DEV_NAME that you'll be able to use in driver when binding driver bus controller

avisconti replied:
After doing this above I'm able to obtain:
#define DT_ST_STM32_I2S_40015000_ST_MP34DT05_40015000_MP34DT05_BUS_NAME            "I2S_5"

Very long BUS_NAME. 
But at the end I should be able to do, in dts_fixup,

#define DT_MP34DT05_MASTER_DEV_NAME 
 DT_ST_STM32_I2S_40015000_ST_MP34DT05_40015000_MP34DT05_BUS_NAME

avisconti replied:
What do you think?
I was expecting a shorter name, like

DT_ST_STM32_I2S_40015000_ST_MP34DT05_BUS_NAME

avisconti replied:
@erwango
CHanged and committed.
Now this define will be generated:

#define DT_ST_STM32_I2S_40015000_ST_MP34DT05_40015000_MP34DT05_BUS_NAME "I2S_5"

It is used to define DT_MP34DT05_MASTER_DEV_NAME which will be used in mp34dt05 driver.

erwango replied:
Indeed, we should anyway a name that has no address reference, so we'll need a fixup indeed.
Then, there might be an issue in generation script and 'ST_MP34DT05_40015000_MP34DT05_BUS_NAME' looks indeed strange. Maybe this is coming from the fact there is no reg associated to the mic node. I'll check tomorrow.

Meanwhile, can you push using the fixup you proposed?

avisconti replied:
Yes, I already did it yesterday.
I followed your instructions and added

 parent:
      bus: i2s

in microphone yaml file.

erwango replied:
So I confirm that string 'DT_ST_STM32_I2S_40015000_ST_MP34DT05_40015000_MP34DT05_BUS_NAME' is due to lack of reg in 96b_argonkey mp34dt05 node.
With a reg we get something like DT_ST_STM32_I2S_40015000_ST_MP34DT05_1_BUS_NAME.

If no reg is as expected we need to hack extract_dts_include script to take this into account
Do you confirm this is as expected. Isn't it possible to have several mic on the same bus?

avisconti replied:
For 'reg' you mean 'address', like hts221@5f for I2C case?
I'm not a big expert of I2S bus, but as far as I understood there is no 'address' concept in I2S. It is possible indeed to have two microphones attached to same I2S controller (i.e. for audio stereo), but they are seen as one from controller point of view. It uses I2S_WS wire to sample mic1 (WS low) and mic2 (WS hi) alternatevely.

Nevertheless, in this case (stereo with 2 mics) I guess that we need to declare both as subnodes of the same I2S controller, so we probably need to use '0' and '1' like mp34dt05@0 and mp34dt05@1.

Correct me if I'm wrong, but I think that this goes back to #9034 where we had a discussion with address-cell and size.

avisconti committed:
dts/i2s: Extend the I2S dts binding with subnodes capability

This commit introduces the possibility to have multiple
device nodes attached to the same I2S controller. For this
purpose a new i2s-device.yaml description has been introduced
with the a 'reg' property to define the logic number of the
device. For example, if two microphones are attached to the
same I2S port (say 1) to achieve stereo audio, the two microphones
might be described in dts as:

    &i2s1 {
        status = "ok";

        mic@0 {
            compatible = "...";
            reg = <0>;
            label = "...";
        };
        mic@1 {
            compatible = "...";
            reg = <1>;
            label = "...";
        };
    };

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
Resubmitted again taking care of @erwango observations on dts.

In PR #9034 there has been a discussion on introducing #{address,size}-cells for i2s nodes. At that
time we decided to not having it. Actually they are needed as we can having two microphones (for
example) attached to the same I2S controller. So now we may have multiple subnodes to I2S
descriptions, each one with a 'reg' property (say mp34dt05@0). This helps in having  more readable 
generated dts names.
.

avisconti commented:
@erwango 
Are you ok with latest submit?

erwango commented on ext/hal/st/lib/audio/microphone/OpenPDMFilter.c:
```diff
@@ -0,0 +1,314 @@
+* You may not use this file except in compliance with the License.
+* You may obtain a copy of the License at:
+*
+*        http://www.st.com/software_license_agreement_image_v2
```
This license does not fit.
-1 waiting this issue to be solved.

avisconti replied:
Decision has been taken: we will move to Apache license since:
1. There are no patents associated with the OpenPDM2PCM Library.
    It is just a simple FIR filter.
2. The library code has been written by scratch. So the library code
     HAS NOT been derived from any code external to ST.
3. Black Duck tool didn't find any matching with other external piece of code

Header will be like this:

\/*                                                                                                                                                                                             
 \* Copyright (c) 2018 STMicroelectronics
 \*
 \* SPDX-License-Identifier: Apache-2.0
 \*/

avisconti commented:
repushed:
 - changed license for external lib into Apache 2.0
 - microphone driver changed name to be generic for st microphones

erwango commented on drivers/audio/st_dmic_i2s.c:
```diff
@@ -0,0 +1,210 @@
+K_MEM_SLAB_DEFINE(rx_pdm_mem_slab, PDM_BLOCK_MAX_SIZE_BYTES, NUM_RX_BLOCKS, 1);
+
+/* ST_DMIC_I2S private data */
+static struct st_dmic_i2s_pdata {
```
Might be worth setting this as a device data already, if we're to support multi-instance at one point

avisconti replied:
makes sense

erwango commented on drivers/audio/Kconfig.st_dmic_i2s:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig AUDIO_ST_DMIC_I2S
```
And is that compliant with https://github.com/zephyrproject-rtos/zephyr/pull/11854?

avisconti replied:
Well, this driver requires I2S name because does not depend to what ST microphone we use, but how we access it. In this case we use I2S. Same microphone can be attached through different communication and will require a different driver.

Before the driver name was the microphone one (i.e. mp34dt05.c) but I changed it because it does not work if same microphone is attache to other bus. For example the disco_l475_iot1 board has same microphone but we cannot use this driver.

erwango replied:
ok, clear enough

avisconti commented on ext/hal/st/lib/audio/microphone/README:
```diff
@@ -0,0 +1,45 @@
+   20:9952bef19da1
+
+Maintained-by:
+   Arm Mbed OS community
```
OK, my mistake.
But it is not clear to me what to do here. Should I completely remove the 'Maintaned-by' or should I change into something like 'maintained by zephyr community'?

erwango reviewed (changes requested):
License is ok, but commit message requires an update.
Otherwise diver _cfg struct should be moved to _data

avisconti commented:
repushed taking into consideration following stuff:
- correct all ext lib license observations including the commit message
- Moved all the stuff into data structure instead of config structure (which is NULL now)
- changed driver name to mpxxdtyy to better suite the microphone device names
- put the I2S routine in a differnt file. It would be nice to extend this in the future adding SAI support to zephyr.

avisconti commented:
@erwango 
This PR is ready. It would be great if it could fit in the upcoming release.

erwango commented on ext/hal/st/lib/audio/microphone/README:
```diff
@@ -0,0 +1,45 @@
+
+
+Dependencies:
+   This library depends by Zephyr I2S driver and is linked statically.
```
replace I2S by mpxxdtyy

avisconti replied:
oh, I missed it.

erwango reviewed (changes requested):
One minor comment otherwise it's fine

avisconti commented:
@erwango 
Do you have other points?

galak commented:
@dbkinder can you review the docs again.

dbkinder reviewed (dismissed):
still one edit mentioned on line 109 in  ``samples/boards/96b_argonkey/microphone/README.rst``

avisconti commented:
> still one edit mentioned on line 109 in `samples/boards/96b_argonkey/microphone/README.rst`

thanks again!
Fixed it and rebased on latest commit of master.
Now it should be ready to be merge before v1.14 window is closed.

avisconti commented:
Eheheh ... rebased and repushed!
We should be ok finally. Giddap!

MaureenHelm commented on samples/boards/96b_argonkey/microphone/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.13.1)
```
Move this below the license/copyright header

avisconti replied:
ack (did it also for CMakelist.txt in sensor)

MaureenHelm commented on boards/arm/96b_argonkey/96b_argonkey.yaml:
```diff
@@ -15,3 +15,4 @@ supported:
   - rtc
   - spi
   - vl53l0x
+  - i2s
```
Move this up to keep alpha order

avisconti replied:
ack

MaureenHelm commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,158 @@
+        .channel = {
+            .req_num_chan = 1,
+        },
+   };
```
This is quite a lot of code for the README and will likely get out of sync with the actual c file. I think you should remove it.

MaureenHelm commented on samples/boards/96b_argonkey/microphone/README.rst:
```diff
@@ -0,0 +1,158 @@
+-----------------
+
+The ttyUSB0 port must be configured in raw mode to avoid having
+characters 'cooked' out.
```
'cooked'? Is this a well-known term?

avisconti replied:
Yes. Basically TTY drivers has two modes 'raw' vs 'cooked'. In this sample we need raw otherwise some characters may be processed by the driver itself and not sent out, corrupting the audio stream (you will hear a lot of noise).

MaureenHelm commented on samples/boards/96b_argonkey/microphone/sample.yaml:
```diff
@@ -0,0 +1,9 @@
+    harness: sensor
+    platform_whitelist: 96b_argonkey
+    tags: sensors
+    depends_on: i2c i2s gpio
```
Why do you need i2c?

MaureenHelm commented on samples/boards/96b_argonkey/microphone/src/main.c:
```diff
@@ -0,0 +1,192 @@
+#include <led.h>
+#include <i2s.h>
+#include <i2c.h>
+#include <spi.h>
```
Do you need to include i2s, i2c, spi? I don't seem them being used directly in the sample.

avisconti replied:
Correct.
i2c.h was useless even before, while i2s.h and spi.h became useless with this sample
new version which uses dmic i/f.

avisconti replied:
So, I removed them.

avisconti commented:
@MaureenHelm 
Pls let me know if I addressed your requested changes properly.

galak commented:
@dbkinder ping.

MaureenHelm commented:
@dbkinder 	please revisit

avisconti commented:
@dbkinder 
Can you confirm the doc part is fine now?
It would be great to have this change merged in v1.14.

avisconti commented:
Rebased on top of 0f4ec93.

avisconti commented:
#12780 has been merged.
Re-based, re-tested and re-pushed.
(still waiting for @dbkinder ack)
